File: packages/server/backbone/rust/src/websocket.rs
use std::collections::HashMap;
use std::sync::Arc;
use std::time::Duration;

use axum::extract::State;
use axum::extract::ws::{Message, WebSocket, WebSocketUpgrade};
use axum::http::HeaderMap;
use axum::response::IntoResponse;
use tokio::sync::{RwLock, mpsc};
use uuid::Uuid;

use crate::db;
use crate::handlers::AppState;

pub struct WSConsumer {
    pub sender: mpsc::UnboundedSender<Message>,
}

pub struct WSHub {
    pub clients: HashMap<String, WSConsumer>,
}

pub fn new_ws_hub() -> Arc<RwLock<WSHub>> {
    Arc::new(RwLock::new(WSHub {
        clients: HashMap::new(),
    }))
}

pub async fn ws_hub_run(_hub: Arc<RwLock<WSHub>>) {
    loop {
        tokio::time::sleep(Duration::from_secs(3600)).await;
    }
}

pub async fn handle_ws_upgrade(
    ws: WebSocketUpgrade,
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
) -> impl IntoResponse {
    let user_agent = headers
        .get("User-Agent")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("")
        .to_string();

    ws.on_upgrade(move |socket| handle_socket(socket, state, user_agent))
}

async fn handle_socket(mut socket: WebSocket, state: Arc<AppState>, user_agent: String) {
    let id = Uuid::new_v4().to_string().replace('-', "");
    let path = "/ws".to_string();
    let remote_addr = String::new();

    let (msg_sender, mut msg_receiver) = mpsc::unbounded_channel::<Message>();

    let consumer = WSConsumer {
        sender: msg_sender,
    };

    state
        .ws_hub
        .write()
        .await
        .clients
        .insert(id.clone(), consumer);
    if let Ok(conn) = state.db.get().await {
        db::insert_ws_connection(&conn, &id, &remote_addr, &path, &user_agent).ok();
    }

    let mut interval = tokio::time::interval(Duration::from_secs(30));
    loop {
        tokio::select! {
            msg = socket.recv() => {
                match msg {
                    Some(Ok(Message::Text(text))) => {
                        if let Ok(conn) = state.db.get().await {
                            db::insert_ws_message(&conn, &id, "received", &text).ok();
                        }
                    }
                    Some(Ok(Message::Close(_))) | None => break,
                    _ => {}
                }
            }
            msg = msg_receiver.recv() => {
                match msg {
                    Some(Message::Close(_)) => {
                        let _ = socket.send(Message::Close(None)).await;
                        break;
                    }
                    Some(msg) => {
                        if socket.send(msg).await.is_err() {
                            break;
                        }
                    }
                    None => break,
                }
            }
            _ = interval.tick() => {
                if socket.send(Message::Ping(vec![].into())).await.is_err() {
                    break;
                }
            }
        }
    }

    state.ws_hub.write().await.clients.remove(&id);
    if let Ok(conn) = state.db.get().await {
        db::update_ws_disconnect(&conn, &id).ok();
    }
}

pub async fn send_to_client(hub: &RwLock<WSHub>, id: &str, message: Message) -> bool {
    let hub = hub.read().await;
    if let Some(client) = hub.clients.get(id) {
        client.sender.send(message).is_ok()
    } else {
        false
    }
}

pub async fn close_client(hub: &RwLock<WSHub>, id: &str) -> bool {
    let mut hub = hub.write().await;
    if let Some(consumer) = hub.clients.remove(id) {
        let _ = consumer.sender.send(Message::Close(None));
        true
    } else {
        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn new_ws_hub_creates_non_null_hub() {
        let hub = new_ws_hub();
        assert!(hub.try_read().is_ok());
    }

    #[tokio::test]
    async fn send_to_client_unknown_returns_false() {
        let hub = new_ws_hub();
        let result = send_to_client(&hub, "nonexistent", Message::Text("hi".into())).await;
        assert!(!result);
    }

    #[tokio::test]
    async fn close_client_unknown_returns_false() {
        let hub = new_ws_hub();
        let result = close_client(&hub, "nonexistent").await;
        assert!(!result);
    }
}

File: packages/server/backbone/go/src/websocket.go
package main

import (
	"database/sql"
	"log"
	"net/http"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

type WSClient struct {
	ID          string
	Hub         *WSHub
	Conn        *websocket.Conn
	Send        chan []byte
	RemoteAddr  string
	UserAgent   string
	Path        string
	ConnectedAt string
}

type WSHub struct {
	mu         sync.RWMutex
	clients    map[string]*WSClient
	register   chan *WSClient
	unregister chan *WSClient
	broadcast  chan []byte
	db         *sql.DB
}

func NewWSHub(db *sql.DB) *WSHub {
	return &WSHub{
		clients:    make(map[string]*WSClient),
		register:   make(chan *WSClient),
		unregister: make(chan *WSClient),
		broadcast:  make(chan []byte, 256),
		db:         db,
	}
}

func (h *WSHub) Run() {
	for {
		select {
		case client := <-h.register:
			h.mu.Lock()
			h.clients[client.ID] = client
			h.mu.Unlock()
			insertWSConnection(h.db, client.ID, client.RemoteAddr, client.Path, client.UserAgent)

		case client := <-h.unregister:
			h.mu.Lock()
			if _, ok := h.clients[client.ID]; ok {
				delete(h.clients, client.ID)
				close(client.Send)
			}
			h.mu.Unlock()
			updateWSDisconnect(h.db, client.ID)

		case message := <-h.broadcast:
			h.mu.RLock()
			for _, client := range h.clients {
				select {
				case client.Send <- message:
				default:
				}
			}
			h.mu.RUnlock()
		}
	}
}

func (h *WSHub) SendToClient(id string, message []byte) bool {
	h.mu.RLock()
	client, ok := h.clients[id]
	h.mu.RUnlock()
	if !ok {
		return false
	}
	select {
	case client.Send <- message:
		return true
	default:
		return false
	}
}

func (h *WSHub) CloseClient(id string) {
	h.mu.RLock()
	client, ok := h.clients[id]
	h.mu.RUnlock()
	if ok {
		h.unregister <- client
		client.Conn.Close()
	}
}

func (h *WSHub) ActiveCount() int {
	h.mu.RLock()
	defer h.mu.RUnlock()
	return len(h.clients)
}

var wsUpgrader = websocket.Upgrader{
	CheckOrigin:     func(r *http.Request) bool { return true },
	ReadBufferSize:  1024,
	WriteBufferSize: 1024,
}

func (h *WSHub) ServeWS(w http.ResponseWriter, r *http.Request) {
	conn, err := wsUpgrader.Upgrade(w, r, nil)
	if err != nil {
		log.Printf("ws upgrade: %v", err)
		return
	}

	id := generateID()
	client := &WSClient{
		ID:          id,
		Hub:         h,
		Conn:        conn,
		Send:        make(chan []byte, 256),
		RemoteAddr:  r.RemoteAddr,
		UserAgent:   r.UserAgent(),
		Path:        r.URL.Path,
		ConnectedAt: time.Now().UTC().Format(time.RFC3339),
	}

	h.register <- client
	go client.writePump()
	go client.readPump()
}

func (c *WSClient) readPump() {
	defer func() {
		c.Hub.unregister <- c
		c.Conn.Close()
	}()
	for {
		_, message, err := c.Conn.ReadMessage()
		if err != nil {
			break
		}
		insertWSMessage(c.Hub.db, c.ID, "received", string(message))
	}
}

func (c *WSClient) writePump() {
	ticker := time.NewTicker(30 * time.Second)
	defer func() {
		ticker.Stop()
		c.Conn.Close()
	}()
	for {
		select {
		case message, ok := <-c.Send:
			if !ok {
				c.Conn.WriteMessage(websocket.CloseMessage, []byte{})
				return
			}
			if err := c.Conn.WriteMessage(websocket.TextMessage, message); err != nil {
				return
			}
		case <-ticker.C:
			if err := c.Conn.WriteMessage(websocket.PingMessage, nil); err != nil {
				return
			}
		}
	}
}
